### PR TITLE
feat: Add Duration/Instant/LocalTime generators

### LINKED ISSUE #6: DateTime Generators
It would be nice to have generators for the most common types of `java.time` api (instant, zone, localDate, localTime, localDateTime, etc.)

It would obviously be available only for the JVM.

### PR DESCRIPTION
**Pull request by alvaroReina**
Resolves #6 

This change is [Reviewable](https://reviewable.io/reviews/jcornaz/kwik/181)

### CODE REVIEW

File: generator/stdlib/src/jvmMain/kotlin/com/github/jcornaz/kwik/generator/stdlib/DateTime.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package com.github.jcornaz.kwik.generator.stdlib
+
+import com.github.jcornaz.kwik.generator.api.Generator
+import java.time.Duration
+import java.time.Instant
+import java.time.LocalTime
+import java.util.*
+
+const val DAY_NANOSECONDS = 86400000000000L
```
Comment (jcornaz) [inline generator/stdlib/src/jvmMain/kotlin/com/github/jcornaz/kwik/generator/stdlib/DateTime.kt:9]: You can make this private ;-)
```suggestion
private const val DAY_NANOSECONDS = 86400000000000L
```

**Comment by jcornaz** [comment]
Hey, thanks for the pull request!

Tell me if you have questions or need help to complete the tests ;-)

**Comment by jcornaz** [comment]
Also we may consider to add the instant corresponding to millisecond `0` as a *sample*.

```kotlin
create { random ->
  Instant.ofEpochMilli(random.nextLong(from = min.toEpochMilli(), until = max.toEpochMilli()))
}.withSamples(Instant.ofEpochMilli(0L))
```


Being a *sample* means it'll have a high probability to be generated, which makes sense because it is kind of an *edge-case* that have higher probability to find errors.

**Comment by alvaroReina** [comment]
Thanks for your valuable feedback. 

Didn't know about the samples, I will try to find more cases where they seem a good fit such as including both min and max values as samples. I have one question regarding your example, What should happen if 0L is not in the generator range?

Regarding the tests, I'm currently looking at the existing generators tests, they're helping me to understand better how it works and what I should take care of. I expect to include better tests in my next update.

**Comment by jcornaz** [comment]
> What should happen if 0L is not in the generator range?

Then it should never be generated. In other words, the sample `0` should not be added.

You may have a look at my approach in 

* the integer generator: https://github.com/jcornaz/kwik/blob/main/generator/stdlib/src/commonMain/kotlin/com/github/jcornaz/kwik/generator/stdlib/Primitives.kt#L20
* this string generator: https://github.com/jcornaz/kwik/blob/main/generator/stdlib/src/commonMain/kotlin/com/github/jcornaz/kwik/generator/stdlib/Strings.kt#L32

**Comment by jcornaz** [comment]
> Didn't know about the samples, I will try to find more cases where they seem a good fit such as including both min and max values as samples. I have one question regarding your example, What should happen if 0L is not in the generator range?

Actually it might be simple an powerfull to simply build upon the long generator.

Like this:

```kotlin
fun Generator.Companion.instants(min: Instant, max: Instant): Generator<Instant> = 
  longs(min.nanoseconds, max.nanoseconds).map { Instant.fromNanoseconds(it) }
```

The advantage is that `longs` already emits `0, -1, 1, min and max` as samples.

**Comment by alvaroReina** [comment]
Initially I had an overflow problem when using .toEpochMillis() on Instant.MAX and Instant.MIN due to the imposibility of representing them with milliseconds precision using Long type.

```kotlin
return create { random ->
    Instant.ofEpochSecond(
        random.nextLong(from = min.epochSecond, until = max.epochSecond),
        random.nextLong(from = 0, until = MAX_NANOSECONDS)
    )
}.withSamples(samples)
```
[stdlib/DateTime.kt#31](https://github.com/jcornaz/kwik/pull/181/files#diff-eb85f2388ffe011b86a28d38ec4e04b1R31)

I fixed it by applying this workaround - using the seconds with nanoadjustment constructor - it solves the previous problem but now I'm facing the following:

- If both min and max are in epochSecond 0 it fails due to invalid range: `Random range is empty: [0, 0).` 
- I'm potentially generating values outside the given range by not taking care of the initial and ending nanoseconds range in the nanoAdjustment argument.

I'll be working on this problem before adding the remaining types, feel free to suggest changes or missing tests that I haven't thought of.

**Comment by jcornaz** [comment]
Hi, thanks for your work.


I'll try to have a look at it tonight or tomorrow and see if I can help ;-)

**Comment by alvaroReina** [comment]
Hi, I managed to solve that problem, I'm going to start working on the missing types (LocalDate, LocalDateTime, Period).

Duration, Instant and LocalTime are ready to be reviewed and I'll let you know when the whole PR is ready, thanks!

**Comment by jcornaz** [comment]
I had a quick look and it looks very good so far.

Let me know when I can give the final review and merge ;-)

P.S: You don't have to support every time types. I could merge generators for Duration, Instant and LocalTime and let LocalDate, LocalDateTime, Period be implemented in another pull request.

**Comment by alvaroReina** [comment]
Thanks, that sounds good. In that case I prefer to narrow down the scope of this PR to just these 3 types. 

It is ready to be reviewed.

**Pull request merged**